### PR TITLE
Represent function pointers in mir-constants as a Value instead of Item

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb

### CODE REVIEW

File: src/librustc_trans/mir/constant.rs [hunk 1]
```diff
@@ -477,8 +480,8 @@ impl<'a, 'tcx> MirConstContext<'a, 'tcx> {
                 let ty = self.monomorphize(&constant.ty);
                 match constant.literal.clone() {
                     mir::Literal::Item { def_id, substs } => {
-                        // Shortcut for zero-sized types, including function item
-                        // types, which would not work with MirConstContext.
+                        // Shortcut for zero-sized types
+                        // which would not work with MirConstContext.
```
Comment (eddyb) [inline src/librustc_trans/mir/constant.rs]: I think this means it's a function-specific hack that can be removed.

File: src/librustc_trans/mir/constant.rs [hunk 2]
```diff
@@ -927,8 +930,8 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
         let ty = self.monomorphize(&constant.ty);
         let result = match constant.literal.clone() {
             mir::Literal::Item { def_id, substs } => {
-                // Shortcut for zero-sized types, including function item
-                // types, which would not work with MirConstContext.
+                // Shortcut for zero-sized types
```
Comment (eddyb) [inline src/librustc_trans/mir/constant.rs]: Same here.

**Review by eddyb — CHANGES_REQUESTED** [review]
LGTM except the hacks can be removed now!

**Comment by oli-obk** [comment]
Gotta run, so I can't test the changes locally. let's see what travis says.

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
You have a merge commit, please rebase properly.

**Comment by oli-obk** [comment]
since the MIR-shim changes I can't figure out how to make this work.

**Comment by oli-obk** [comment]
I applied the fixes as discussed in IRC. It works now.

**Pull request merged**